crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs

```diff
@@ -60,42 +286,705 @@ declare_lint_rule! {
+        self.0
+    }
+
+    pub fn is_forbidden(&self) -> bool {
+        !self.0
+    }
+
+    pub fn reason(&self) -> ImportRestrictionCause {
+        self.1
+    }
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    Deserialize,
+    biome_deserialize_macros::Deserializable,
+    Eq,
+    PartialEq,
+    Serialize,
+)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields, default)]
+pub struct CustomRestrictedImportOptions {
+    message: String,
+    import_names: Box<[Box<str>]>,
+    allow_import_names: Box<[Box<str>]>,
+}
+
+impl CustomRestrictedImportOptions {
+    pub fn has_import_name_patterns(&self) -> bool {
+        !self.import_names.is_empty() || !self.allow_import_names.is_empty()
+    }
+
+    fn is_import_allowed(&self, imported_name: &str) -> ImportRestrictionStatus {
+        if !self.allow_import_names.is_empty() {
+            // Deny all imports except for the names specified in allow_import_names
+            let is_allowed = self
+                .allow_import_names
+                .iter()
+                .any(|name| &**name == imported_name);
+
+            ImportRestrictionStatus(is_allowed, ImportRestrictionCause::AllowImportNames)
+        } else if !self.import_names.is_empty() {
+            // Allow all imports except for the names specified in import_names
+            let is_forbidden = self
+                .import_names
+                .iter()
+                .any(|name| &**name == imported_name);
+
+            ImportRestrictionStatus(!is_forbidden, ImportRestrictionCause::ImportNames)
+        } else {
+            // Deny all imports from this module
+            ImportRestrictionStatus(false, ImportRestrictionCause::ImportSource)
+        }
+    }
+
+    fn get_message_for_restriction(
+        &self,
+        import_source: &str,
+        imported_name: &str,
+        reason: ImportRestrictionCause,
+    ) -> String {
+        if !self.message.is_empty() {
+            self.message.clone()
+        } else {
+            match reason {
+                ImportRestrictionCause::ImportSource => {
+                    format!("Do not import '{import_source}'.")
+                }
+                ImportRestrictionCause::ImportNames | ImportRestrictionCause::AllowImportNames => {
+                    if imported_name == RestrictedImportVisitor::BARE_IMPORT_ALIAS {
+                        format!("Do not import '{import_source}' through a side-effect import.")
+                    } else {
+                        format!("Do not import '{imported_name}' from '{import_source}'.")
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(untagged)]
+pub enum CustomRestrictedImport {
+    Plain(String),
+    WithOptions(CustomRestrictedImportOptions),
+}
```

Comment:
User-facing options must be documented in doc comments. The comments will become part of the `description` field of the JSON schema, which will be shown to the users in their editors.